File: src/main/java/com/jozufozu/exnihiloomnia/common/entity/EntityThrownStone.java
package com.jozufozu.exnihiloomnia.common.entity;

import com.jozufozu.exnihiloomnia.common.items.ExNihiloItems;
import net.minecraft.block.BlockState;
import net.minecraft.block.SoundType;
import net.minecraft.block.material.Material;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.passive.IronGolemEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.projectile.ThrowableEntity;
import net.minecraft.util.DamageSource;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.SoundEvent;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.world.World;

public class EntityThrownStone extends ThrowableEntity
{
    public EntityThrownStone(World worldIn)
    {
        super(worldIn);
    }
    
    public EntityThrownStone(World worldIn, EntityLivingBase throwerIn)
    {
        super(worldIn, throwerIn);
    }
    
    /**
     * Called when this EntityThrowable hits a block or entity.
     */
    protected void onImpact(RayTraceResult result)
    {
        if (result.getType() == RayTraceResult.Type.ENTITY)
        {
            int i = 4;
    
            if (result.entityHit instanceof IronGolemEntity)
            {
                i = 0;
            }
    
            result.entityHit.attackEntityFrom(DamageSource.causeThrownDamage(this, this.getThrower()), (float) i);
        }
    
        if (!this.world.isRemote)
        {
            if (result.getType() == RayTraceResult.Type.BLOCK)
            {
                BlockPos blockPos = result.getBlockPos();
                BlockState blockState = this.world.getBlockState(blockPos);
    
                //We want it to break glass, but some stuff (barrels) shouldn't

                if (getMotion().lengthSquared() >= 1.2 && blockState.getMaterial() == Material.GLASS && blockState.getBlockHardness(world, blockPos) <= 0.6f && !blockState.hasTileEntity())
                {
                    this.world.destroyBlock(blockPos, false);
                    this.motionX *= 0.8f;
                    this.motionZ *= 0.8f;
                    
                    return;
                }
                
                SoundType soundType = blockState.getBlock().getSoundType(blockState, this.world, blockPos, this);
                SoundEvent hitSound = soundType.getHitSound();
                this.world.playSound(null, result.hitVec.x, result.hitVec.y, result.hitVec.z, hitSound, SoundCategory.PLAYERS, 0.6f, soundType.getPitch() + this.world.rand.nextFloat() * 0.4f);
            }
            
            if (world.rand.nextFloat() <= 0.5f && !(this.getThrower() instanceof PlayerEntity && ((PlayerEntity) this.getThrower()).isCreative()))
                this.dropItem(ExNihiloItems.STONE, 1);
            
            this.world.setEntityState(this, (byte) 3);
            this.remove();
        }
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/common/entity/EntityThrownStone.kt
package com.jozufozu.exnihiloomnia.common.entity

import com.jozufozu.exnihiloomnia.common.ExNihiloTags
import com.jozufozu.exnihiloomnia.common.items.ExNihiloItems
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.passive.IronGolemEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.projectile.ThrowableEntity
import net.minecraft.util.DamageSource
import net.minecraft.util.SoundCategory
import net.minecraft.util.math.BlockRayTraceResult
import net.minecraft.util.math.EntityRayTraceResult
import net.minecraft.util.math.RayTraceResult
import net.minecraft.world.World

class EntityThrownStone : ThrowableEntity {
    constructor(worldIn: World) : super(worldIn) {}

    constructor(worldIn: World, throwerIn: LivingEntity) : super(worldIn, throwerIn) {}

    /**
     * Called when this EntityThrowable hits a block or entity.
     */
    override fun onImpact(result: RayTraceResult) {
        if (result.type == RayTraceResult.Type.ENTITY && result is EntityRayTraceResult) {
            var i = 4

            if (result.entity is IronGolemEntity) {
                i = 0
            }

            result.entity.attackEntityFrom(DamageSource.causeThrownDamage(this, this.thrower), i.toFloat())
        }

        if (!this.world.isRemote) {
            if (result.type == RayTraceResult.Type.BLOCK && result is BlockRayTraceResult) {
                val blockPos = result.pos
                val blockState = this.world.getBlockState(blockPos)

                //We want it to break glass, but some stuff (barrels) shouldn't

                if (motion.lengthSquared() >= 1.2 && ExNihiloTags.Blocks.THROWN_STONE_BREAKING.contains(blockState.block)) {
                    this.world.destroyBlock(blockPos, false)

                    setMotion(motion.x * 0.8, motion.y, motion.z * 0.8)

                    return
                }

                val soundType = blockState.block.getSoundType(blockState, this.world, blockPos, this)
                val hitSound = soundType.hitSound
                this.world.playSound(null, result.hitVec.x, result.hitVec.y, result.hitVec.z, hitSound, SoundCategory.PLAYERS, 0.6f, soundType.getPitch() + this.world.rand.nextFloat() * 0.4f)
            }

            if (world.rand.nextFloat() <= 0.5f && !(this.thrower is PlayerEntity && (this.thrower as PlayerEntity).isCreative))
                this.entityDropItem(ExNihiloItems.STONE)

            this.world.setEntityState(this, 3)
            this.remove()
        }
    }
}